app/services/content_embed_service.rb

```diff
@@ -0,0 +1,83 @@
+class ContentEmbedService
+  include ApplicationHelper
+
+  SUPPORTED_DOCUMENT_TYPES = %w[contact].freeze
+  UUID_REGEX = /([0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})/
+  EMBED_REGEX = /({{embed:(#{SUPPORTED_DOCUMENT_TYPES.join('|')}):#{UUID_REGEX}}})/
+
+  class EmbeddedEdition
+    attr_reader :embed_code, :document_type, :content_id
+
+    def initialize(embed_code:, document_type:, content_id:)
+      @embed_code = embed_code
+      @document_type = document_type
+      @content_id = content_id
+    end
+
+    def edition
+      @edition ||= Edition.with_document.find_by(
+        state: "published",
+        content_store: "live",
+        document_type: @document_type,
+        documents: { content_id: @content_id },
+      )
+    end
+  end
+
+  def initialize(body)
+    @body = body
+  end
+
+  def embedded_editions
+    @embedded_editions ||= @body.scan(EMBED_REGEX).map do |match|
+      EmbeddedEdition.new(
+        embed_code: match[0],
+        document_type: match[1],
+        content_id: match[2],
+      )
+    end
+  end
+
+  def fetch_links
+    embedded_editions.group_by(&:document_type).transform_values do |items|
+      items.map(&:content_id)
+    end
+  end
+
+  def render
+    embedded_editions.each do |embedded_edition|
+      replace_embed_code_with_content(embedded_edition.embed_code, embedded_edition.edition)
+    end
+
+    @body
+  end
+
+private
+
+  def find_edition(document_type, content_id)
+    Edition.with_document.find_by(
+      state: "published",
+      content_store: "live",
+      document_type:,
+      documents: { content_id: },
+    )
+    ## TODO: What do we do if the edition can't be found?
```

Comment:
Some kind of error I guess...

We might have to think about the UX here too, as I suspect this processing will be happening inside a sidekiq job, so it might not be possible to communicate back to the publishing app that the user has referred to a UUID that doesn't match up with an edition in the right state for the right document type.

We'll also have to think about edge cases like "if someone unpublishes or withdraws an object that's embedded in other editions, what should happen when those editions are republished?". I feel like we'll want some kind of equivalent to `ON DELETE RESTRICT` in SQL, where it's not possible to unpublish / otherwise remove an object if its embedded in any published / draft editions.